homeassistant/components/squeezebox_player/media_player.py

```diff
@@ -0,0 +1,218 @@
+"""MediaPlayer platform for Squeezebox integration."""
+from __future__ import annotations
+
+import asyncio
+
+from aioslimproto.client import PlayerState, SlimClient
+from aioslimproto.const import EventType, SlimEvent
+from aioslimproto.server import SlimServer
+
+from homeassistant.components import media_source
+from homeassistant.components.media_player import (
+    BrowseMedia,
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_IDLE, STATE_OFF, STATE_PAUSED, STATE_PLAYING
+from homeassistant.core import EventOrigin, HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.dt import utcnow
+
+from .const import DEFAULT_NAME, DOMAIN, PLAYER_EVENT
+
+SUPPORTED_FEATURES = (
+    MediaPlayerEntityFeature.PAUSE
+    | MediaPlayerEntityFeature.VOLUME_SET
+    | MediaPlayerEntityFeature.STOP
+    | MediaPlayerEntityFeature.TURN_ON
+    | MediaPlayerEntityFeature.TURN_OFF
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.PLAY_MEDIA
+    | MediaPlayerEntityFeature.VOLUME_MUTE
+    | MediaPlayerEntityFeature.BROWSE_MEDIA
+)
+
+STATE_MAPPING = {
+    PlayerState.IDLE: STATE_IDLE,
+    PlayerState.PLAYING: STATE_PLAYING,
+    PlayerState.PAUSED: STATE_PAUSED,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Squeezebox MediaPlayer(s) from Config Entry."""
+    slimserver: SlimServer = hass.data[DOMAIN]
+    added_ids = set()
+
+    async def async_add_player(player: SlimClient) -> None:
+        """Add MediaPlayerEntity from SlimClient."""
+        if player.player_id in added_ids:
+            return
+        # we delay adding the player a small bit because the player name may be received
+        # just a bit after connect. This way we can create a device reg entry with the correct name
+        for _ in range(10):
+            if player.player_id not in player.name:
+                break
+            await asyncio.sleep(0.1)
+        added_ids.add(player.player_id)
+        async_add_entities([SqueezeboxPlayer(slimserver, player)])
+
+    async def on_slim_event(event: SlimEvent):
+        """Handle player added/connected event."""
+        player = slimserver.get_player(event.player_id)
+        await async_add_player(player)
+
+    # register listener for new players
+    config_entry.async_on_unload(
+        slimserver.subscribe(on_slim_event, EventType.PLAYER_CONNECTED)
+    )
+
+    # add all current items in controller
+    for player in slimserver.players:
+        await async_add_player(player)
```

Comment:
Use `asyncio.gather` since there is a sleep in here.

```suggestion
    await asyncio.gather(*(async_add_player(player) for for player in slimserver.players))
```

Reply:
small detail is that the sleep is only reached when a new player connects (first message is hello, second optional message is its name)

Reply:
Actually we don't have to wait?
```suggestion
    for player in slimserver.players:
        hass.async_create_task(async_add_player(player))
```

Reply:
well in case of a race condition we do (player connects just at the exact same time we're initializing).
But create task is fine here. we just want to make sure we have the initial state